src/lib.rs

```diff
@@ -8,35 +8,246 @@ mod transaction;
+
+        let message0_bob = bob0.next_message();
+        transport
+            .send_message(Message::CreateMessage0(message0_bob))
+            .await?;
+
+        let message0_alice = match transport.receive_message().await? {
+            Message::CreateMessage0(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+        let bob1 = bob0.receive(message0_alice, wallet).await.unwrap();
+
+        let message1_bob = bob1.next_message();
+        transport
+            .send_message(Message::CreateMessage1(message1_bob))
+            .await?;
+
+        let message1_alice = match transport.receive_message().await? {
+            Message::CreateMessage1(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+        let bob2 = bob1.receive(message1_alice).unwrap();
+
+        let message2_bob = bob2.next_message();
+        transport
+            .send_message(Message::CreateMessage2(message2_bob))
+            .await?;
+
+        let message2_alice = match transport.receive_message().await? {
+            Message::CreateMessage2(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+        let bob3 = bob2.receive(message2_alice).unwrap();
+
+        Self::create(transport, wallet, bob3).await
+    }
+
+    async fn create<W, T>(
+        transport: &mut T,
+        wallet: &W,
+        party3: create::Party3,
+    ) -> anyhow::Result<Self>
+    where
+        W: BuildFundingPSBT + SignFundingPSBT + BroadcastSignedTransaction,
+        T: SendMessage + ReceiveMessage,
+    {
+        let message3_self = party3.next_message();
+        transport
+            .send_message(Message::CreateMessage3(message3_self))
+            .await?;
+
+        let message3_other = match transport.receive_message().await? {
+            Message::CreateMessage3(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+        let party4 = party3.receive(message3_other).unwrap();
+
+        let message4_self = party4.next_message();
+        transport
+            .send_message(Message::CreateMessage4(message4_self))
+            .await?;
+
+        let message4_other = match transport.receive_message().await? {
+            Message::CreateMessage4(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+        let party5 = party4.receive(message4_other).unwrap();
+
+        let message5_self = party5.next_message(wallet).await.unwrap();
+        transport
+            .send_message(Message::CreateMessage5(message5_self))
+            .await?;
+        let message5_other = match transport.receive_message().await? {
+            Message::CreateMessage5(message) => message,
+            _ => anyhow::bail!("wrong message"),
+        };
+
+        let (channel, transaction) = party5.receive(message5_other, wallet).await.unwrap();
+
+        wallet.broadcast_signed_transaction(transaction).await?;
```

Comment:
I think it's fine to leave it, maybe add a todo but realistically, if the broadcast fails then we should still return the channel because the other party is likely to have done the broadcast (except if the transaction is invalid).

We need to return the channel so the consumer can still try to close it.

Reply:
Yep. I think if broadcasting fails we need to look in the blockchain for `TX_f`'s txid. If that also fails we should still return the channel struct like I think you're saying: so that we can still manipulate it if the channel is eventually opened by the other party or even so that we can attempt to open it ourselves again.

The return type of this function could be `Result<State>` where State could be something like

```rust
enum State {
    Open(Channel),
    Primed(Channel, Transaction),
}
```

The error variant of the result would capture errors that happen during the create protocol, whereas the enum would capture whether the channel was opened or not, with `Primed` giving you enough to recover/continue.

Alternatively, you could model it with a `Result<Result<Channel, BroadcastError>, ProtocolError>` and let the `BroadcastError` hold the `Channel` and `signed_TX_f` inside.

Reply:
But yeah, we can do this in a separate PR.

Reply:
Sounds good, maybe open an issue to track?